### PR TITLE
When required text codecs aren't available on the system, proxy the conversion over to GDAL's API

### LINKED ISSUE #36871: encoding CP852 missing in windows instalation
**There is no possibility to set encoding CP 852 to layer in windows instalation. **
Code CP 852 is not in the list of encodings. But is available in ubuntu instalation. If I have shapefile with this encoding defined in cpg file the encoding pre-set to System. QGIS seems like ignore cpg file. There is no more possibility so set this by settings ->datasource->ignore shapefile encoding declaration in this version.

**QGIS and OS versions**
QGIS 3.10.6 on win 10 and ubuntu 18.04

### PR DESCRIPTION
**Pull request by nyalldawson**
Resolves missing text codecs like CP852 on windows builds. These
were previously available, but then Qt upstream dropped the ICU
library from their windows builds, and accordingly a whole bunch
of older text codecs are no longer available by default on the
windows builds.

Fixes #36871

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault I'd love your eyes on this one!

**Comment by agiudiceandrea** [comment]
Does this fix also allow to **create** a shapefile with a CP852 encoding (or another one not available in Qt for Windows)?

File: src/core/qgsogrproxytextcodec.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+{
+  static QStringList codecs;
+  static std::once_flag initialized;
+  std::call_once( initialized, [&]( )
+  {
+    // there's likely others that are supported by GDAL, but we're primarily concerned here
+    // with codecs used by the shapefile driver, and which are no longer supported on the
+    // windows Qt builds (due to removal of ICU support in windows Qt builds)
+    // see https://github.com/qgis/QGIS/issues/36871
+    for ( int i = 437; i <= 950; ++i )
```
Comment (rouault) [inline src/core/qgsogrproxytextcodec.cpp:91]: don't know if that matters in the usage, but not all codes in that range will be valid. On Unix, the list could probably be retrieved from iconv API. On Windows, WideCharToMultiByte() API is used

Comment (nyalldawson) [reply src/core/qgsogrproxytextcodec.cpp:91]: Any clue why Travis is segfaulting with the new test? I can't reproduce locally, and valgrind is happy...

File: src/core/qgsogrproxytextcodec.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  return result;
+}
+
+QByteArray QgsOgrProxyTextCodec::convertFromUnicode( const QChar *unicode, int length, ConverterState * ) const
+{
+  if ( !unicode )
+    return QByteArray();
+
+  const QString src = QString( unicode, length );
+  char *res = CPLRecode( src.toLocal8Bit().constData(), CPL_ENC_UTF8, mName.constData() );
```
Comment (rouault) [inline src/core/qgsogrproxytextcodec.cpp:50]: shouldn't that rather be something like (untested) ?
```suggestion
  char *res = CPLRecode( src.toUtf8().constData(), CPL_ENC_UTF8, mName.constData() );
```

File: src/core/qgsogrproxytextcodec.cpp (hunk 1 above)
Comment (rouault) [reply src/core/qgsogrproxytextcodec.cpp:91]: I could reproduce it in a Docker container mimicking Travis
The issue is the one pointed at https://doc.qt.io/qt-5/qtextcodec.html#QTextCodec-1

I've confirmed the following change in TestQgsOgrUtils::textCodec() fixes the crash
```
+  // The QTextCodec should always be constructed on the heap. Qt takes ownership and will delete it when the application terminates.
+  QgsOgrProxyTextCodec* codec = new QgsOgrProxyTextCodec( "CP852" );
+  QCOMPARE( codec->toUnicode( codec->fromUnicode( "abcŐ" ) ), QStringLiteral( "abcŐ" ) );
+  QCOMPARE( codec->toUnicode( codec->fromUnicode( "" ) ), QString() );
+  // cppcheck-suppress memleak
```

That also means that qgsvectordataprovider.h must be changed to not free mProxyCodec

**Comment by nyalldawson** [comment]
Let's give this a shot now...

**Comment by i-s-o** [comment]
In addition to the codepage problem, the lack of ICU library has at least two other consequences in some languages/locales:
* Incorrect capitalization, e.g.: https://en.wikipedia.org/wiki/Dotted_and_dotless_I
  `tr_locale = QLocale(QLocale.Turkish)`
  `tr_locale.toUpper("Diyarbakır") # => DIYARBAKIR (wrong, should be DİYARBAKIR)`
* Incorrect collation, e.g.: https://github.com/qgis/QGIS/issues/30573

Is there any chance of compiling Qt with ICU for QGIS distributions? I will search the Web to figure out what it'll take to compile Qt but I don't think I will be to make any contributions in this area with my level of understanding on how compilers work.

**Comment by i-s-o** [comment]
> lack of ICU library has at least two other consequences

@jef-n could you provide some insight on whether/how to include ICU in Qt for QGIS releases?

**Comment by jef-n** [comment]
@i-s-o patience.  the next round of packages have icu support

**Comment by i-s-o** [comment]
@jef-n that's great news, thanks!

When you say "next round", is it with 3.16+Qt 5.11, or with 3.18+Qt 5.15? I am just curious about the timing, no urgency: this issue only came up a year ago while helping a friend. I handled the missing CP857 by using ogr2ogr, and the string case change by installing PyICU, and writing the upper case function in Python.

**Comment by jef-n** [comment]
There's only one Qt in OSGeo4W (not counting the qt4 cruft left) - currently 5.11 and next 5.15.  And all QGIS' will make use of that once the switch is done (ie. 3.10.x and 3.16 and above).

**Pull request merged**